[ESIMD] Implement stateless memory accesses enforcement (#6287)
intel/llvm: sycl/doc/UsersManual.md

    NOTE: This flag is currently only supported with the CUDA and HIP targets.

> **`-f[no]-sycl-esimd-force-stateless-mem`** [EXPERIMENTAL]

Author: Fixed.

    Enforces usage of stateless memory accesses and automatic conversion of
    "stateful" memory accesses via SYCL accessors to "stateless" accesses
    within ESIMD (Explicit SIMD) kernels by compiler.

>     This option disables the intrinsics and methods accepting SYCL accessors

Reviewer: It disables in -f or -fno- form? Can you clarify here?

Author: -f option disables those intrinsics that cannot be converted automatically. I updated the text. Thank you.

    or "surface-index" and that cannot be automatically converted to their
    "stateless" equivalents.

    NOTE: "Stateful" accesses are those that use SYCL accessors or a pair
    of "surface-index" + 32-bit byte-offset and use specific memory access
    data port messages to read/write/fetch.
    "Stateless" accesses are those that use memory locations represented
    with virtual memory addresses such as USM pointers.

    For some of usages using "stateless" memory may be beneficial as it
    does not have  the limit of 4Gb per surface.
    Also, some of Intel GPUs or GPU run-time/drivers may support only
    "stateless" memory accesses.

> 

Reviewer: Please add information about default value.

Author: Added. Thank you.

    NOTE: This flag is currently only supported with the CUDA and HIP targets.

**`-f[no-]sycl-esimd-force-stateless-mem`** [EXPERIMENTAL]

>     Enforces usage of stateless memory accesses and automatic conversion of
>     "stateful" memory accesses via SYCL accessors to "stateless" accesses
>     within ESIMD (Explicit SIMD) kernels by compiler.

Reviewer 2: ```suggestion
    Enforce stateless memory access and automatic conversion of "stateful" 
    memory access via SYCL accessors to "stateless" accesses within ESIMD
    (Explicit SIMD) kernels.
```

Author: I updated this section here: fc687f617ef5a730bcd9436e4d6cdd5ea33354ef

>     This option disables the intrinsics and methods accepting SYCL accessors

Reviewer 2: ```suggestion
    -fsycl-esimd-force-stateless-mem disables the intrinsics and methods accepting SYCL accessors
```

Author: Fixed. Thank you. fc687f617ef5a730bcd9436e4d6cdd5ea33354ef

>     or "surface-index" and that cannot be automatically converted to their

Reviewer 2: ```suggestion
    or "surface-index" which cannot be automatically converted to their
```

I'm not sure if I misunderstood intent here. So please ignore my suggestion if I did.

Author: Fixed. fc687f617ef5a730bcd9436e4d6cdd5ea33354ef

    converted to their "stateless" equivalents.

>     The stateless memory access enforcement is disabled by default.
> 
>     -fno-sycl-esimd-force-stateless-mem is used to tell compiler not to
>     enforce usage of stateless memory accesses.

Reviewer: ```suggestion
    -fno-sycl-esimd-force-stateless-mem is used to tell compiler not to
    enforce usage of stateless memory accesses. This is default behavior.
```
nit: looks a bit clearer to me this way.

Author: Ok, Thank you. Fixed here: 11d9c6d5f6e687f1a1f438143b7a59ca4ca3e06a
